torchdata/datapipes/iter/util/combining.py

```diff
@@ -72,3 +68,52 @@ def __iter__(self):
 
     def __len__(self):
         return len(self.source_datapipe)
+
+
+def tuple_merge(item, item_from_map):
+    return (item, item_from_map)
+
+
+@functional_datapipe("zip_with_map")
+class MapZipperIterDataPipe(IterDataPipe):
+    r""" :class:`MapZipperIterDataPipe`.
+
+    IterDataPipe that joins the items from the source IterDataPipe with items from a MapDataPipe. The
+    matching is done by the key function, which maps an item from source IterDataPipe to
+    a key that exists in MapDataPipe. The return value is created by the merge function, which returns
+    a tuple of the two items by default.
+
+    Args:
+        source_iterdatapipe: IterDataPipe from which items are yield and will be combined with an item from map_datapipe
+        map_datapipe: MapDataPipe that takes a key from key_fn, and returns an item
+        key_fn: Function that maps each item from source_iterdatapipe to a key that exists in map_datapipe
+        merge_fn: Function that combines the item from source_iterdatapipe and the item from map_datapipe,
+            by default a tuple is created
+    """
+
+    def __init__(
+        self,
+        source_iterdatapipe: IterDataPipe,
+        map_datapipe: MapDataPipe,
+        key_fn: Callable,
+        merge_fn: Callable = tuple_merge,
+    ):
+        self.source_iterdatapipe = source_iterdatapipe
+        self.map_datapipe = map_datapipe
```

Comment:
Could we add a check if it's `MapDataPipe`?

Reply:
Done